src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter_Statics.cs

```diff
@@ -99,6 +106,42 @@ public static void CompileMethod(WebAssemblyCodegenCompilation compilation, WebA
             methodCodeNodeNeedingCode.SetDependencies(ilImporter.GetDependencies());
         }
 
+        private static void EmitNativeToManagedThunk(WebAssemblyCodegenCompilation compilation, MethodDesc method, string nativeName, LLVMValueRef managedFunction)
+        {
+            LLVMTypeRef[] llvmParams = new LLVMTypeRef[method.Signature.Length];
+            for (int i = 0; i < llvmParams.Length; i++)
+            {
+                llvmParams[i] = GetLLVMTypeForTypeDesc(method.Signature[i]);
+            }
+
+            LLVMTypeRef thunkSig = LLVM.FunctionType(GetLLVMTypeForTypeDesc(method.Signature.ReturnType), llvmParams, false);
+            LLVMValueRef thunkFunc = LLVM.AddFunction(compilation.Module, nativeName, thunkSig);
+            LLVMBasicBlockRef block = LLVM.AppendBasicBlock(thunkFunc, "Block0");
+            LLVMBuilderRef builder = LLVM.CreateBuilder();
+            LLVM.PositionBuilderAtEnd(builder, block);
+
+            LLVMValueRef shadowStack = LLVM.BuildArrayAlloca(builder, LLVM.Int8Type(), LLVM.ConstInt(LLVM.Int32Type(), 1000000, LLVMMisc.False), "shadowStack");
```

Comment:
I don't know that this needs to be addressed right now but this appears to be allocating a new stack any time we get called by native code. Once we start working on Exceptions and GC it seems like this would need to be at the very least linked into the (reverse) pinvoke transition frame. Maybe even pick up the shadow stack where we left off in a scenario something like this
NativeMain -> Transition -> ManagedMain -> PInvoke -> NativeCallableManagedFunctionPointer

Reply:
You're right... that was a bit hacky. I've switched to storing a pointer to the top of the shadow stack in a thread static that gets set in main and p/invoke transitions and read here. I can't actually test it yet since we don't have any code that will call a RuntimeImport though.